### PR TITLE
Add support for Python 3.11

### LINKED ISSUE #9691: Python 3.11 support
Python 3.11 has been released 🎉 Historically it has taken some time for our dependencies to add support for new Python versions and for those changes to show up in releases. Opening this issue for tracking purposes 

cc @charlesbluca in case this is something you might be interested in

### PR DESCRIPTION
**Pull request by graingert**
- [x] Closes #9691
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by graingert** [comment]
getting a segfault here: https://github.com/dask/dask/actions/runs/3638328422/jobs/6140402545#step:7:11195 looks like https://github.com/pandas-dev/pandas/issues/50105

**Review by jrbourbeau — COMMENTED** [review]
Nice detective work @graingert. Sounds like we can safely `xfail` the unhappy hdf-related tests on Python 3.11

**Comment by graingert** [comment]
we can't xfail them as they segfault - will have to skip them for now

**Comment by jrbourbeau** [comment]
👍

**Comment by graingert** [comment]
depends on https://github.com/dask/distributed/pull/7249

File: continuous_integration/environment-3.11.yaml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  - pytest-timeout
+  - pytest-xdist
+  - moto
+  - flask
+  - fastparquet>=0.8.0
+  - h5py
+  - pytables
+  # - zarr
+  # `tiledb-py=0.17.5` lead to strange seg faults in CI, However 0.18 is needed for 3.11
+  # We should unpin when possible.
```
Comment (crusaderky) [inline continuous_integration/environment-3.11.yaml:26]:
```suggestion
```

File: continuous_integration/environment-3.11.yaml [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+  # https://github.com/dask/dask/pull/9569
+  - tiledb-py
+  # - pyspark
+  - tiledb>=2.5.0
+  - xarray
+  - fsspec
+  - sqlalchemy>=1.4.0
+  # - pyarrow  needs 10+ for python3.11, conda-forge only has v9
```
Comment (crusaderky) [inline continuous_integration/environment-3.11.yaml:34]:
```suggestion
  - pyarrow>=10
```
it's out now

**Review by jrbourbeau — COMMENTED** [review]
Restarting failed builds as it looks like  the linux Python 3.11 build stalled for some reason

File: continuous_integration/environment-3.11.yaml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  - python-graphviz
+  - python-xxhash
+  - mmh3
+  - jinja2
+  - pip
+  # The nightly pyarrow / arrow-cpp packages currently don't install with latest
+  # protobuf / abseil, see https://github.com/dask/dask/issues/9449
+  - libprotobuf=3.19
+  - pip:
+    - git+https://github.com/graingert/distributed@python-311
```
Comment (graingert) [inline continuous_integration/environment-3.11.yaml:76]:
```suggestion
    - git+https://github.com/dask/distributed
```

File: continuous_integration/environment-3.11.yaml [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  - python-xxhash
+  - mmh3
+  - jinja2
+  - pip
+  # The nightly pyarrow / arrow-cpp packages currently don't install with latest
+  # protobuf / abseil, see https://github.com/dask/dask/issues/9449
+  - libprotobuf=3.19
+  - pip:
+    - git+https://github.com/dask/distributed
+    - pyarrow  # pyarrow on conda-forge is 9.0 needs 10+ for python3.11
```
Comment (graingert) [inline continuous_integration/environment-3.11.yaml:76]:
```suggestion
```

**Comment by jakirkham** [comment]
Should this still be a draft?

**Comment by graingert** [comment]
Hmm pyarrow isn't working due to

> Encountered problems while solving:
>    - package pyarrow-10.0.1-py310he8a299c_0_cuda requires libarrow 10.0.1 ha344585_0_cuda, but none of the providers can be installed

**Comment by jakirkham** [comment]
Would just pin to Arrow 9. Likely it will require work to upgrade to a newer version

**Comment by jakirkham** [comment]
There should be a Python 3.11 package for Arrow 9 already

It is just Arrow 10 came out somewhat recently and there are some fairly significant changes in it that should improve things overall. Just will take time to fully adapt to. So that's what we are seeing here

**Comment by crusaderky** [comment]
@graingert commented in a previous version of the PR that pyarrow 10 is necessary for python 3.11

**Comment by jakirkham** [comment]
That doesn't sound right. Arrow 9 was definitely built for Python 3.11 (for [example]( https://github.com/conda-forge/arrow-cpp-feedstock/blob/6f2e688ac79a9c123c2721abee0221074b7dac67/.azure-pipelines/azure-pipelines-linux.yml#L35 )). If that doesn't work, would be very interested to know.

**Review by jrbourbeau — COMMENTED** [review]
Okay, looks like the issue was due to the `libprotobuf` pin. That's only needed for the `pyarrow`-nightly build and has since been moved to a different file anyways. After removing the `libprotobuf` pin, I was able to get a solved Python 3.11 environment locally on my mac. Let's see if it solves in CI.

**Comment by jrbourbeau** [comment]
Looks like that did the trick

**Review by crusaderky — APPROVED** [review]

**Review by jrbourbeau — COMMENTED** [review]
The Linux 3.11 build stalled for some reason, so I just retriggered CI to see if the stall happens consistently or not. If it doesn't, I think this is good to go

**Comment by mrocklin** [comment]
I'll be glad to see this get in.  Thank you everyone who helped here.  I'm excited for Python 3.11.

**Comment by jrbourbeau** [comment]
Okay, looks like the hanging Linux Python 3.11 build passed (🎉 ) but we're seeing `dask/tests/test_multiprocessing.py::test_custom_context_used_python3_posix` fail on macOS Python 3.11. We're also seeing it fail consistently. I'm not able to reproduce locally when I run that test in isolation. Looking into it now... Would certainly welcome thoughts from others too : )

**Comment by jrbourbeau** [comment]
Alright, looks like `tiledb` may have been the source of the issue. At least locally when I run the `tiledb`-related tests alongside `test_custom_context_used_python3_posix` with `pytest dask -k "test_custom_context_used_python3_posix or tiledb"` I'm able to reproduce the error. I've temporarily commented `tiledb` out of the Python 3.11 CI environment file. 🤞 that fixes things

**Review by jrbourbeau — APPROVED** [review]
Thanks @graingert @crusaderky @jakirkham!

**Pull request merged**